### DISCUSSION TITLE
“Run in Terminal…” (and bash= arguments) not working

### DISCUSSION THREAD

**Discussion by lukestein** [swiftbar/SwiftBar#189] [Q&A] [answered]
I am a bit confounded and hope someone may have an idea for me. My plugins run and update just fine in the menubar. However, if I choose “Run in Terminal…” they do *not* run in the Terminal. The Terminal app will be opened (if it was closed) or brought to the front, but my script is not run.

The same is true if I choose a menu item containing a `bash=` argument… it opens Terminal, but doesn’t run anything.

I don’t know if this is a failure of AppleScript control, or something else. I have given SwiftBar the only permissions it seems to have asked for (i.e., Automation > System Events)
<img width="780" alt="image" src="https://user-images.githubusercontent.com/44452336/115488518-a379eb80-a228-11eb-9fb3-d8ee38b12ce3.png">

I’ve been running Swiftbar 1.2.3 on Big Sur, but also just tried the 1.3.0 beta to see if it helped—no luck. Any advice much appreciated!

**Comment by melonamin** [comment, marked as answer]
- Do you see anything in the Terminal at all? Or it is just blank?
- What Terminal are you using, standard or iTerm?

**Reply by lukestein** [reply]
Thanks for your help! I am using the standard terminal. Swiftbar opens the Terminal app (if it is unopened), or brings it to the foreground, but doesn’t run anything—I just get my standard shell prompt.

Perhaps this 30-sec video will show exactly what I mean. I show “Run in Terminal…”, running the script manually from the shell, and selecting from the menu (which I think should run `ssh pizero` in the terminal).

https://user-images.githubusercontent.com/44452336/115597508-56892a00-a2a7-11eb-9627-738767344cb2.mp4

**Reply by melonamin** [reply]
Weird... can you share a plugin you're using for testing this? Just to eliminate any uncertainty in testing on my end

**Reply by lukestein** [reply]
The plugin I demoed is this (just a slightly modified version of [ssh.sh](https://github.com/matryer/xbar-plugins/blob/master/Network/ssh.sh) from the plugin directory): 
```bash
#!/bin/bash
temp=$(ssh pi "vcgencmd measure_temp | cut -c6-9")
echo "$temp℃"
echo "---"
awk  '/^Host / && !/\*/ {print $2" | bash=ssh param1="$2}' ~/.ssh/config
```

I get this behavior from all plugins though (including straight from the directory).

Not sure whether relevant, but I did just notice that my terminal defaults to `zsh` rather than `bash` (I believe this is now the Mac default?).

If there’s anything else I can try to help run/report on my end, let me know; I’ll do my best.

**Reply by lukestein** [reply]
Also, can confirm that `| href=` works fine; there's only a problem with `| bash=`

**Reply by melonamin** [reply]
Looks like the Automation permissions is the problem, as an example here is how the pref pane looks on my machine:

<img width="348" alt="CleanShot 2021-04-21 at 15 13 33@2x" src="https://user-images.githubusercontent.com/222100/115608418-39a72380-a2b4-11eb-8728-c2bbb92e28ea.png">

SwiftBar should request the permission on the first need, i.e. when you first time clicked on "Run In Terminal". I'm not sure why you don't have it, myself suggestion would be to try it on another macOS user or  wipe SwiftBar completely and reinstall.


This is what you should delete in to addition to SwiftBar itself:

* "~/Library/Application Scripts/com.ameba.SwiftBar-LaunchAtLoginHelper"
* "~/Library/Caches/com.ameba.SwiftBar"
* "~/Library/Containers/com.ameba.SwiftBar-LaunchAtLoginHelper"
* "~/Library/Preferences/com.ameba.SwiftBar.plist"


Sorry, I don't have a better guess at this moment.

**Reply by lukestein** [reply]
So. You got me on the right track and now… **success**. Thank you!

I had already tried reinstalling the App, but hadn't deleted these caches and settings. By itself, deleting them didn't work. Then I did some digging around. It [seems that the permissions are stored centrally](https://bitsplitting.org/2018/07/11/reauthorizing-automation-in-mojave/) [very useful link] and you can't reset a single app's requests; you have to reset all of them 🤦). The automation requests reset with `tccutil reset AppleEvents`, but it seems that perhaps SwiftBar wasn't able to ask again for more automation privileges unless it *also* had accessibility privileges. I tried a bunch of combinations, and here's what worked (partly documenting in case anyone else runs into this):

1. Quit and delete SwiftBar app from Applications folder
2. Delete preferences, scripts, and caches as you suggested: `rm -rf ~/Library/Application\ Scripts/com.ameba.SwiftBar-LaunchAtLoginHelper & rm -rf ~/Library/Caches/com.ameba.SwiftBar & rm -rf ~/Library/Containers/com.ameba.SwiftBar-LaunchAtLoginHelper & rm -rf ~/Library/Preferences/com.ameba.SwiftBar.plist`
3. Restart computer
4. Run `tccutil reset All`. Note this will require re-agreeing to *all* app privilege requests.
5. Restart computer *again*
6. Install SwiftBar and accept privilege requests.

**Reply by melonamin** [reply]
OMG! Glad you've fixed it and thank you for the write up, I'll direct people here in the future :)
